File: answers/leetcode/String to Integer (atoi).py
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """

        if '' == str:
            return 0
        i = 0
        while i < len(str) and str[i].isspace():
            i += 1
        sign = 1
        if i < len(str) and '-' == str[i]:
            sign = -1
        if i < len(str) and ('-' == str[i] or '+' == str[i]):
            i += 1
        INT_MAX = 2147483647
        INT_MIN =  - INT_MAX - 1
        integer = 0
        while i < len(str) and str[i].isdigit():
            digit = int(str[i])
            if integer > INT_MAX / 10 or integer * 10 + digit > INT_MAX:
                return INT_MAX if 1 == sign else INT_MIN
            else:
                integer = integer * 10 + digit;
            i += 1
        return sign * integer

File: answers/leetcode/String to Integer (atoi)/String to Integer (atoi).java
class Solution {
    public int myAtoi(String str) {
        boolean start = false;
        boolean negative = false;
        int number = 0;
        for (int i = 0; i < str.length(); i++) {
            char c = str.charAt(i);
            if ((c == ' ' || c == '\t') && !start) {
                continue;
            } else if (c == '+' && !start) {
                start = true;
            } else if (c == '-' && !start) {
                negative = true;
                start = true;
            } else if (Character.isDigit(c)) {
                start = true;
                int digit = Character.getNumericValue(c);
                if (number > Integer.MAX_VALUE / 10
                    || (number == Integer.MAX_VALUE / 10 && digit > Integer.MAX_VALUE % 10)) {
                    return negative ? Integer.MIN_VALUE : Integer.MAX_VALUE;
                }
                number = number * 10 + digit;
            } else {
                break;
            }
        }
        return number * (negative ? -1 : 1);
    }
}